### ISSUE TITLE
New API to set each property of the dialog separately

### ISSUE THREAD

**Issue by btzy** [btzy/nativefiledialog-extended#92]
**EDIT: A versioned struct will be used instead of this, implemented in #128.**

The current API looks like:
```
nfdresult_t res = NFD_OpenDialogN(outPath, filterList, filterCount, "/path/to/default/folder");
```
This interface is inextensible.  It is impossible to add new parameters without breaking API and ABI, so it is not possible to provide some improvements in a non-breaking manner.  They include:
- Setting a custom title on a dialog (#63) (this feature is also required by Linux applications wishing to provide non-English translations of the text "Open File" etc, since unlikely Windows and MacOS, GTK does not provide a default title).
- Setting a parent window handle (`HWND` on Windows, parent_window string on Portal, etc) (required to properly fix #90).

Adding a new API that sets each property separately will allow support for additional parameters, and also make it more future-proof, like the following:
```
NFD_OpenFileDialog* dialog = NFD_OpenFileDialog_Create();
NFD_OpenFileDialog_SetFilterList(dialog, filterList, filterCount);
NFD_OpenFileDialog_SetDefaultFolder(dialog, "/path/to/default/folder");
nfdresult_t res = NFD_OpenFileDialog_Show(dialog, outPath);
```

I'm not decided on this and am not sure that this is the best way yet - feel free to leave comments.

The current API will be retained, and it will call the new API under the hood.

**Comment by krupkat** [comment]
You probably thought about this, but there is also the option of renaming the new extended function, e.g. to NFD_OpenDialogN_v2 and then call the new function from the old one.

I'm curious - in your example, if you change the NFD_OpenFileDialog struct, won't the ABI still break?

**Comment by btzy** [comment]
The main reason for naming it `NFD_OpenFileDialog` instead of `NFD_OpenDialog` is the name clash with the existing function, but I'm open to using other names that might be less verbose.

**Comment by krupkat** [comment]
Thanks for taking the time to explain, it makes sense! The renaming method isn't sustainable long term, I agree.

**Comment by btzy** [comment]
Another interesting possibility is to put the following in `nfd.h`:

```c
typedef struct {
    Filter* filter_list = NULL;
    size_t filter_count = 0;
    const char* title = NULL;
    const char* current_folder = NULL;
    bool modal = true;
    const char* parent_window = NULL;
    /* more fields might be added here in the future so this struct should not be passed across an ABI boundary */
} NFD_OpenDialogProperties;

/* This is the new function to call to show an open file dialog */
nfd_result NFD_OpenDialogWithProperties(const char** outPath, NFD_OpenDialogProperties props) {
    // This calls some library functions in some order supported by the library (the order and the way it is called depends on the platform)
    // E.g. on Linux we might need to pass the parent window when constructing the dialog
    void* handle = NFD_OpenDialog_Impl_New(props.parent_window);
    if (filter_count) NFD_OpenDialog_Impl_SetFilterList(handle, props.filter_list, props.filter_count);
    if (title) NFD_OpenDialog_Impl_SetTitle(handle, props.title);
    if (current_folder)NFD_OpenDialog_Impl_SetCurrentFolder(handle, props.current_folder);
    // E.g. `result` and `modal` are passed together here due to the shape of the platform native API
    return NFD_OpenDialog_Impl_Show(handle, outPath, props.modal);
}

/* APIs to set individual properties - while these are stable, users should not use them directly because the functions may need to be called in a specific order!  These functions may vary between backends.  Use NFD_OpenDialogWithProperties() instead in user code.  */
void* NFD_OpenDialog_Impl_New(const char* parent_window);
void NFD_OpenDialog_Impl_SetFilterList(void* handle, Filter* filter_list, size_t filter_count);
void NFD_OpenDialog_Impl_SetTitle(void* handle, const char* title);
void NFD_OpenDialog_Impl_SetCurrentFolder(void* handle, const char* current_folder);
bool NFD_OpenDialog_Impl_Show(void* handle, const char** result, bool modal);
```

The user will then write the following code to open the file dialog (all the properties are optional, just set what you need):

```c
char* result;
if (NFD_OpenDialogWithProperties(&result, {
        .filter_list = my_list,
        .filter_count = 4,
        .parent_window = get_native_window()}) == NFD_OKAY) {
    printf("Processing file %s", result);
    free(result);
}
```

As long as the caller does not rely on the size of the `NFD_OpenDialogProperties` struct, they won't need to modify their code to compile with a newer version of NFDe (i.e. no API break).  If NFDe is compiled as a shared library, it is also possible to update the NFDe shared library without recompiling the calling code (i.e. no ABI break).

Programs that depend on NFDe should not pass the `NFD_OpenDialogProperties` across an ABI boundary, but this seems very unlikely for most projects.  Probably the only projects that might consider about putting `NFD_OpenDialogProperties` in a public interface are libraries that bundle NFDe with other platform abstraction libraries, but putting a warning about that in the NFDe documentation/README should probably suffice.

However, I have never seen this style of interface in other open source libraries.

**Comment by btzy** [comment]
The current plan is to pass a struct together with a version number, and when new fields are added to the end of the struct, the version number will be bumped.  It is implemented in #128.
